### ISSUE TITLE
Massive (7000+ms) lag spikes on world save/autosave

### ISSUE THREAD

**Issue by justicesays** [Mithion/ArsMagica2#1002]
Been having some lag issues on the server, and have been forced to turn off autosave and confine saving to hourly windows where several seconds of lag while it saves is less of an issue and can be avoided for critical periods. This has its own problems with unloading world chunks, rollbacks etc. so is less than ideal.
I just installed sampler and took an application snapshot during the world save spike.
The culprit appears to be am2.power.PowerNodeCache.onWorldSave() ->SaveNBTToFile 7304ms (98.7%)
I don't have a hugely complex set up, me and one other guy have a few devices, some obelisks, two dark aurems, about 15ish celestial aurems and some relays chains, a couple of longish ones to the mob farms for the dark essence.
Not sure if there is something about our setups that is causing it, moving stuff about is pretty much impossible due to the irremovable power break blocks that spawn.

My FML log output looks like this
[18:37:13] [Server thread/DEBUG] [FML/]: Gathering id map for writing to world save world
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 4 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 9 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 11 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 9 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 3 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 8 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 18 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 6 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 3 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 29 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 9 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 6 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 5 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 7 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 13 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 4 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 13 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 4 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:13] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:20] [Server thread/INFO] [sampler/]: Captured 7403 ms lag spike.
[18:37:20] [Server thread/INFO] [sampler/]: Lag spike sample exported to spike-7403-2277719460.nps.
[18:37:20] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:20] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 7 power node entries
[18:37:20] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:20] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 1 power node entries
[18:37:20] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:20] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 13 power node entries
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Saved 2 power node entries
[18:37:21] [Server thread/INFO] [sampler/]: Captured 164 ms lag spike.
[18:37:21] [Server thread/INFO] [sampler/]: Lag spike sample exported to spike-164-2277719675.nps.
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Node list not found.  Checking cache/files for prior data
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded 2 power node entries
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded power data for chunk 2, 28
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 2, 28
[18:37:21] [Server thread/INFO] [sampler/]: Captured 170 ms lag spike.
[18:37:21] [Server thread/INFO] [sampler/]: Lag spike sample exported to spike-170-2277719892.nps.
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Node list not found.  Checking cache/files for prior data
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded 2 power node entries
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded power data for chunk 21, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 21, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Node list not found.  Checking cache/files for prior data
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded 7 power node entries
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded power data for chunk 22, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -27
[18:37:21] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Node list not found.  Checking cache/files for prior data
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded 2 power node entries
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded power data for chunk 19, -25
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 19, -25
[18:37:22] [Server thread/INFO] [sampler/]: Captured 413 ms lag spike.
[18:37:22] [Server thread/INFO] [sampler/]: Lag spike sample exported to spike-413-2277720599.nps.
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Node list not found.  Checking cache/files for prior data
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded 1 power node entries
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded power data for chunk 21, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Node list not found.  Checking cache/files for prior data
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded 13 power node entries
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded power data for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Located Power Node list for chunk 22, -26
[18:37:22] [Server thread/INFO] [sampler/]: Captured 268 ms lag spike.
[18:37:22] [Server thread/INFO] [sampler/]: Lag spike sample exported to spike-268-2277720869.nps.
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Node list not found.  Checking cache/files for prior data
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded 1 power node entries
[18:37:22] [Server thread/TRACE] [FML/]: Ars Magica 2 >> Loaded power data for chunk 22, -28
[18:37:22] [Server thread/INFO] [sampler/]: Captured 143 ms lag spike.
[18:37:22] [Server thread/INFO] [sampler/]: Lag spike sample exported to spike-143-2277721433.nps.

![Uploading am2spike.jpg . . .]()

**Comment by Mithion** [comment]
So, AM2 attempts to save its power data whenever the world does (or when a chunk unloads, which is necessary or you lose the power data)...so there is no autosave specific to AM2 really.

Honestly it looks like chunks are being unloaded/loaded with a great frequency on your server.

What version are you running?  Are you running Bukkit/Cauldron?  It really messes with chunk loading, and AM2's code isn't optimized to work that way.

**Comment by justicesays** [comment]
I'm running
Minecraft 1.7.10
forge-1.7.10-10.13.2.1230
AM2_1.4.0.006.jar
AnimationAPI-1.7.10-1.2.4
Automagy-1.7.10-0.18
ChickenChunks-1.7.1-1.3.4.16-universal
CodeChickenCore-1.7.10-1.0.4.29-universal
extrautilities-1.2.1
fastcraft-1.13
MobiusCore-1.2.3_1.7.10
NewDungeons-0.3(1.7.10)
NotEnoughItems-1.7.10-1.0.3.74-universal
Opis-1.2.3_1.7.10
OptiFine_1.7.10_HD_B4 (client side only)
Thaumcraft-1.7.10-4.2.3.2
twilightforest-1.7.10-2.3.2
Baubles-1.7.10-1.0.1.10
CodeChickenLib-1.7.10-1.1.1.99-universal

No Bukkit or cauldron.
My power setup is in my base which is ( at least partially) chunckloaded via chickenchunks, so probably shouldnt unload at all. although some of it is probably out of range of the chunkloaders.
My mates base is isnt chunkloaded specifically but does seem to have about 8 chunks loaded at all times, probably mirror links to my base.
Total server chunks loaded are typically ~600 loaded, 44 forced chunks , when static in my base, due to having autosave turned off typically these are unloading slower than normal, I turn on autosave every five mins so minecraft can purge down the loaded chunks. However this issue occurs whenever a save-all or autosave happens,I get this whether I'm moving around the world or standing still.
This snapshot was taken about 30 mins after server start, near my base after having just logged in.
Its the second of two snapshots, the first one occurred an autosave earlier (about 2 mins), I left autosave on to capture the debug. The first snapshot shows identical behaviour, 7334 ms (98.4%) in the am2.power.PowerNodeCache.SaveNBTToFile.
Between these two snapshots , I stood still.

**Comment by Mithion** [comment]
So I think a lot of this is that AM2 is doing file I/O when chunks unload, which is never very fast.  I had hoped that it would be a little better, but I rolled the dice and got burned.

I've added a config option to turn off the file I/O on chunk unload, and instead it will be cached.  This will result in a larger memory footprint, and more vulnerability in the event of a crash, but will vastly increase performance when saving.  

The power system will still completely save itself when the world does, but I have also added a config to turn this off too, basically only saving when a world is unloading (usually on server stop).  
This will again sacrifice instability protection for performance.  It will be up to server admins to tune these settings to their liking.
